python/cudf_polars/cudf_polars/experimental/join.py

```diff
@@ -0,0 +1,275 @@
+# SPDX-License-Identifier: Apache-2.0
+"""Parallel Join Logic."""
+
+from __future__ import annotations
+
+import operator
+from functools import reduce
+from typing import TYPE_CHECKING, Any
+
+from cudf_polars.dsl.ir import Join
+from cudf_polars.experimental.base import PartitionInfo, _concat, get_key_name
+from cudf_polars.experimental.dispatch import generate_ir_tasks, lower_ir_node
+from cudf_polars.experimental.shuffle import Shuffle, _partition_dataframe
+
+if TYPE_CHECKING:
+    from collections.abc import MutableMapping
+
+    from cudf_polars.dsl.expr import NamedExpr
+    from cudf_polars.dsl.ir import IR
+    from cudf_polars.experimental.parallel import LowerIRTransformer
+
+
+def _make_hash_join(
+    ir: Join,
+    output_count: int,
+    partition_info: MutableMapping[IR, PartitionInfo],
+    left: IR,
+    right: IR,
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    shuffle_options: dict[str, Any] = {}
+
+    if not (
+        partition_info[left].partitioned_on == ir.left_on
+        and partition_info[left].count == output_count
+    ):
+        left = Shuffle(
+            left.schema,
+            ir.left_on,
+            shuffle_options,
+            left,
+        )
+        partition_info[left] = PartitionInfo(
+            count=output_count,
+            partitioned_on=ir.left_on,
+        )
+
+    if not (
+        partition_info[right].partitioned_on == ir.right_on
+        and partition_info[right].count == output_count
+    ):
+        right = Shuffle(
+            right.schema,
+            ir.right_on,
+            shuffle_options,
+            right,
+        )
+        partition_info[right] = PartitionInfo(
+            count=output_count,
+            partitioned_on=ir.right_on,
+        )
+
+    new_node = ir.reconstruct([left, right])
+
+    partitioned_on: tuple[NamedExpr, ...] = ()
+    if ir.left_on == ir.right_on or (ir.options[0] in ("left", "semi", "anti")):
+        partitioned_on = ir.left_on
+    elif ir.options[0] == "right":  # pragma: no cover
+        partitioned_on = ir.right_on
+    partition_info[new_node] = PartitionInfo(
+        count=output_count,
+        partitioned_on=partitioned_on,
+    )
+
+    return new_node, partition_info
+
+
+def _make_bcast_join(
+    ir: Join,
+    output_count: int,
+    partition_info: MutableMapping[IR, PartitionInfo],
+    left: IR,
+    right: IR,
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    how = ir.options[0]
+    if how != "inner":
+        shuffle_options: dict[str, Any] = {}
+        left_count = partition_info[left].count
+        right_count = partition_info[right].count
+        if left_count >= right_count:
+            right = Shuffle(
+                right.schema,
+                ir.right_on,
+                shuffle_options,
+                right,
+            )
+            partition_info[right] = PartitionInfo(
+                count=right_count,
+                partitioned_on=ir.right_on,
+            )
```

Comment:
I see this logic in a few places. Should we consolidate in a helper function?

Reply:
Yeah, makes sense - I centralized this in a `_maybe_shuffle_frame` utility.